File: apps/relay-go/pkg/engine/penalties.go
package engine

import (
	"fmt"

	"github.com/prof-faustus/cardtable/relay-go/pkg/types"
)

// ClassifyResult names the round outcome and whether the third card
// was actually consulted (penalty outcomes resolve before examining
// the third card).
type ClassifyResult struct {
	Outcome          types.SettlementOutcome
	ThirdCardChecked bool
}

// ClassifyInBetweenRound returns the outcome class for an In-Between
// round given the three visible cards. Per spec/card-protocol.md and
// the test vectors:
//
//   - if the first two cards share a rank-ord -> equal_penalty
//   - else if they are consecutive (rank-ord difference == 1) -> consecutive_penalty
//   - else if the third card's rank-ord is strictly between the two ->
//     win
//   - else -> loss
func ClassifyInBetweenRound(visible []types.RevealedCard) (ClassifyResult, error) {
	if len(visible) != 3 {
		return ClassifyResult{}, fmt.Errorf("ClassifyInBetweenRound: expected 3 visible cards, got %d", len(visible))
	}
	a := visible[0].Ordinal % 13
	b := visible[1].Ordinal % 13
	if a == b {
		return ClassifyResult{Outcome: types.OutcomeEqualPenalty, ThirdCardChecked: false}, nil
	}
	diff := a - b
	if diff < 0 {
		diff = -diff
	}
	if diff == 1 {
		return ClassifyResult{Outcome: types.OutcomeConsecutivePenalty, ThirdCardChecked: false}, nil
	}
	c := visible[2].Ordinal % 13
	low, high := a, b
	if a > b {
		low, high = b, a
	}
	if c > low && c < high {
		return ClassifyResult{Outcome: types.OutcomeWin, ThirdCardChecked: true}, nil
	}
	return ClassifyResult{Outcome: types.OutcomeLoss, ThirdCardChecked: true}, nil
}

// ValueTransfer is the signed delta of the active player's balance
// and the pot for one settlement. Conservation: PlayerDelta + PotDelta
// == 0.
type ValueTransfer struct {
	PlayerDelta int64
	PotDelta    int64
}

// ComputeValueTransfer is the pure (outcome, bet, rules) -> deltas
// function. Use int64 deltas so a loss can be a negative number;
// callers convert back to Satoshis at the call site.
func ComputeValueTransfer(outcome types.SettlementOutcome, bet types.Satoshis, rules types.SettlementRules) (ValueTransfer, error) {
	switch outcome {
	case types.OutcomeWin:
		win := int64(bet) * int64(rules.InBetweenWinMultiplier)
		return ValueTransfer{PlayerDelta: win, PotDelta: -win}, nil
	case types.OutcomeLoss:
		loss := int64(bet) * int64(rules.InBetweenLossMultiplier)
		return ValueTransfer{PlayerDelta: -loss, PotDelta: loss}, nil
	case types.OutcomePass:
		return ValueTransfer{PlayerDelta: 0, PotDelta: 0}, nil
	case types.OutcomeConsecutivePenalty:
		p := int64(rules.ConsecutiveCardsPenalty)
		return ValueTransfer{PlayerDelta: -p, PotDelta: p}, nil
	case types.OutcomeEqualPenalty:
		p := int64(rules.EqualCardsPenalty)
		return ValueTransfer{PlayerDelta: -p, PotDelta: p}, nil
	default:
		return ValueTransfer{}, fmt.Errorf("ComputeValueTransfer: unknown outcome %q", outcome)
	}
}

File: packages/state-engine/src/penalties.ts
/**
 * In-Between settlement and penalty logic.
 *
 * Pure functions over (visible cards, bet, rules). Three classes:
 *   - WIN: third card strictly between visible cards (open interval).
 *   - LOSS: third card outside the open interval.
 *   - CONSECUTIVE_PENALTY: visible cards are consecutive (rank difference == 1).
 *   - EQUAL_PENALTY: visible cards have equal rank.
 *
 * The `_PENALTY` outcomes resolve **before** examining the third card:
 * if the visible cards are consecutive or equal, the active player pays a
 * fixed penalty into the pot regardless of the third card.
 */

import type { RevealedCard, Satoshis, SettlementRules } from '@cardtable/protocol-types';
import { asSatoshis } from '@cardtable/protocol-types';
import type { SettlementOutcome } from '@cardtable/protocol-types';

/** Compute the settlement outcome class for an In-Between round. */
export function classifyInBetweenRound(
  visibleCards: readonly RevealedCard[],
): { outcome: SettlementOutcome; thirdCardChecked: boolean } {
  if (visibleCards.length !== 3) {
    throw new Error(
      `classifyInBetweenRound: expected exactly 3 visible cards, got ${visibleCards.length}`,
    );
  }
  const a = visibleCards[0];
  const b = visibleCards[1];
  const c = visibleCards[2];
  if (a === undefined || b === undefined || c === undefined) {
    throw new Error('classifyInBetweenRound: indexing failure');
  }
  const aOrd = a.ordinal % 13;
  const bOrd = b.ordinal % 13;
  if (aOrd === bOrd) {
    return { outcome: 'equal_penalty', thirdCardChecked: false };
  }
  if (Math.abs(aOrd - bOrd) === 1) {
    return { outcome: 'consecutive_penalty', thirdCardChecked: false };
  }
  const cOrd = c.ordinal % 13;
  const low = Math.min(aOrd, bOrd);
  const high = Math.max(aOrd, bOrd);
  if (cOrd > low && cOrd < high) {
    return { outcome: 'win', thirdCardChecked: true };
  }
  return { outcome: 'loss', thirdCardChecked: true };
}

/**
 * Compute the value transfer between the active player and the pot for a
 * given outcome class, bet amount, and rule set.
 *
 * Returns `{ playerDelta, potDelta }` where the player's balance changes
 * by `playerDelta` and the pot changes by `potDelta`. By conservation
 * `playerDelta + potDelta == 0` always.
 */
export function computeValueTransfer(
  outcome: SettlementOutcome,
  bet: Satoshis,
  rules: SettlementRules,
): { playerDelta: number; potDelta: number } {
  switch (outcome) {
    case 'win': {
      // Player wins their bet × win_multiplier from the pot.
      const win = bet * rules.in_between_win_multiplier;
      return { playerDelta: win, potDelta: -win };
    }
    case 'loss': {
      const loss = bet * rules.in_between_loss_multiplier;
      return { playerDelta: -loss, potDelta: loss };
    }
    case 'pass': {
      return { playerDelta: 0, potDelta: 0 };
    }
    case 'consecutive_penalty': {
      const p = rules.consecutive_cards_penalty;
      return { playerDelta: -p, potDelta: p };
    }
    case 'equal_penalty': {
      const p = rules.equal_cards_penalty;
      return { playerDelta: -p, potDelta: p };
    }
    default: {
      const _exhaustive: never = outcome;
      throw new Error(`computeValueTransfer: non-exhaustive outcome ${String(_exhaustive)}`);
    }
  }
}

/** Helper: clamp to Satoshis with a sanity check that the value is non-negative. */
export function clampSat(n: number): Satoshis {
  if (!Number.isSafeInteger(n) || n < 0) {
    throw new Error(`clampSat: result is not a non-negative safe integer: ${n}`);
  }
  return asSatoshis(n);
}
